### PR TITLE
Login

### PR DESCRIPTION
**Pull request by 845968074**

### CODE REVIEW

**Comment by wyb1995** [comment]
:+1:

**Comment by ZhuoYueYue** [comment]
:+1:

**Comment by lidan-summer** [comment]
:+1:

File: app/api/api.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import express from 'express';
+import sessionsApi from './sessions.js';
+const router = express.Router();
+router.use('/sessions', sessionsApi);
+export default router;
```
Comment (freewind) [inline app/api/api.js:5]: /app/api/index.js

File: app/api/sessions.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+const router = express.Router();
+
+router.post('/', function (req, res, next) {
+  const userId = req.body.userId;
+  const password = req.body.password;
+  if (_.isEmpty(userId) || _.isEmpty(password)) {
+    return res.status(400).send({httpCode: 400, message: "数据不能为空"});
+  }
+  else {
+    request.get('https://api.xiyoumobile.com/xiyoulibv2/user/login')
```
Comment (freewind) [inline app/api/sessions.js:16]: 这一大段，抽到/app/services/xiyou/login.js

```
export default function(callback) {
  request.get('https://api.xiyoumobile.com/xiyoulibv2/user/login')
     .send({username: userId, password: password})
     .end((err, response) => {
         if(err) return callback(err);

          var session = respones.body.Detail;
          const loginSuccessfully =  session === 'ACCOUNT_ERROR';
         callback(null, loginSuccessfully);
     }
}
```

File: app/api/sessions.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      .send({username: userId, password: password})
+      .end((err, respones) => {
+        if (err) return next(err);
+        var session = respones.body.Detail;
```
Comment (freewind) [inline app/api/sessions.js:20]: detail

File: app/api/sessions.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        if (session === 'ACCOUNT_ERROR') {
+          /* window.location.href = "index.html";*/
```
Comment (freewind) [inline app/api/sessions.js:22]: delete

File: app/api/sessions.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          return res.status(401).send({httpCode: 401, message: "用户名或密码有误，登录失败"});
```
Comment (freewind) [inline app/api/sessions.js:23]:
```
 return res.status(401).send("用户名或密码有误，登录失败");
```

File: app/api/sessions.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        }
+        else {
+          // window.location.href = "main.html?session=" + Session;
+          /*   window.location.href = "/#/personalInfoPage";*/
```
Comment (freewind) [inline app/api/sessions.js:27]: delete

File: app/api/sessions.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    request.get('https://api.xiyoumobile.com/xiyoulibv2/user/login')
+      .send({username: userId, password: password})
+      .end((err, respones) => {
+        if (err) return next(err);
+        var session = respones.body.Detail;
+        if (session === 'ACCOUNT_ERROR') {
+          /* window.location.href = "index.html";*/
+          return res.status(401).send({httpCode: 401, message: "用户名或密码有误，登录失败"});
+        }
+        else {
```
Comment (freewind) [inline app/api/sessions.js:25]: 不需要这个else

File: app/api/sessions.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+          // window.location.href = "main.html?session=" + Session;
+          /*   window.location.href = "/#/personalInfoPage";*/
+          User.findOne({userId: userId}, function (err, user) {
+            if (err) return next(err);
+            if (user == null) {
+              res.status(201).send({httpCode: 201, message: "SUCCESS", newUser: true})
+            }
+            else {
```
Comment (freewind) [inline app/api/sessions.js:33]: remove else

File: app/api/sessions.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+          /* window.location.href = "index.html";*/
+          return res.status(401).send({httpCode: 401, message: "用户名或密码有误，登录失败"});
+        }
+        else {
+          // window.location.href = "main.html?session=" + Session;
+          /*   window.location.href = "/#/personalInfoPage";*/
+          User.findOne({userId: userId}, function (err, user) {
+            if (err) return next(err);
+            if (user == null) {
+              res.status(201).send({httpCode: 201, message: "SUCCESS", newUser: true})
```
Comment (freewind) [inline app/api/sessions.js:31]: res.status(201).send({message: "SUCCESS", newUser: true})

File: app/api/sessions.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            }
+            else {
+              res.status(201).send({httpCode: 201, message: "SUCCESS", newUser: false})
```
Comment (freewind) [inline app/api/sessions.js:34]: send({message: "SUCCESS", newUser: false})

File: app/api/sessions.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      });
+  }
+});
+
+router.get('/', function (req, res, next) {
+  User.find((err, users) => {
+    if (err) return next(err);
+    res.json(users);
+  });
+});
```
Comment (freewind) [inline app/api/sessions.js:47]: /app/api/users.js

File: app/mongodb/db.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import mongoose from'mongoose';
+module.exports = {
+  connect: function () {
+    mongoose.connect('mongodb://localhost/login');
```
Comment (freewind) [inline app/mongodb/db.js:4]: db name?

File: app/mongodb/user.js [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import mongoose from 'mongoose';
+const Schema = mongoose.Schema;
+
+const User = new Schema({
+  userId: String,
+  password: String,
+  tel: String,
+  email: String
+});
+const Users = mongoose.model('users', User);
```
Comment (freewind) [inline app/mongodb/user.js:10]: User

File: app/server.js [hunk 14]
```diff
@@ -22,11 +25,14 @@ app.use(webpackHotMiddleware(compiler, {
 }));
 
 app.use(express.static('./public'));
+app.use("/api", loginApi);
+/*
+ app.post('/login',mongodb.login);*/
+/*app.post('/personal',mongodb.modify);*/
```
Comment (freewind) [inline app/server.js:31]: delete

File: app/server.js [hunk 15]
```diff
@@ -3,10 +3,13 @@ import webpackConfig from '../webpack.config';
 import webpackDevMiddleware from 'webpack-dev-middleware';
 import webpackHotMiddleware from 'webpack-hot-middleware';
 import express from 'express';
-
+import loginApi from "./api/api.js";
```
Comment (freewind) [inline app/server.js:6]: import apiRouter

File: app/server.js [hunk 16]
```diff
@@ -22,11 +25,14 @@ app.use(webpackHotMiddleware(compiler, {
 }));
 
 app.use(express.static('./public'));
+app.use("/api", loginApi);
```
Comment (freewind) [inline app/server.js:28]: apiRouter

File: app/server.js (hunk 14 above)
Comment (freewind) [inline app/server.js:31]: delete

File: package.json [hunk 17]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "newSchoolRunner",
+  "name": "es6-project-demo",
```
Comment (freewind) [inline package.json:2]: ???

File: public/entry.jsx [hunk 18]
```diff
@@ -1,18 +1,23 @@
...
 import React from 'react';
 import ReactDOM from 'react-dom';
+require('jquery');
+require("bootstrap-webpack");
 import $ from 'jquery';
 
+const route = <Router history={hashHistory}>
+  <Route path="/" component={App}>
+    <IndexRedirect to='/api/sessions'/>
+{/*    <Route path='/api/sessions' component={LoginPage}/>*/}
```
Comment (freewind) [inline public/entry.jsx:11]: ????

File: public/entry.jsx [hunk 19]
```diff
@@ -1,18 +1,23 @@
...
+</Router>;
+
 ReactDOM.render(
-  <Hello />,
+  route,
   document.getElementById("content")
 );
-
-// use jquery
 console.log($('#content').text());
```
Comment (freewind) [inline public/entry.jsx:19]: delete this line

File: spec/dummy-spec.js [hunk 20]
```diff
@@ -2,6 +2,6 @@
 
 describe('dummy', () => {
   it('1+1 == 2', () => {
-    expect(1+1).toEqual(2);
+    expect(1 + 1).toEqual(2);
   })
-});
\ No newline at end of file
+});
```
Comment (freewind) [inline spec/dummy-spec.js:7]: delete this file

File: spec/sessions-spec.js [hunk 21]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+import request from 'supertest';
+import app from '../app/server';
+import finish from './finish';
+import User from '../app/mongodb/user';
```
Comment (freewind) [inline spec/sessions-spec.js:5]: import {User} from './schema';

File: spec/sessions-spec.js [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+
+describe('sessions-spec', () => {
+  beforeEach((done)=> {
+    User.find().remove(finish(done));
+  });
+
+  it('post true id and password sessions', (done) => {
```
Comment (freewind) [inline spec/sessions-spec.js:12]: 用中文

File: spec/sessions-spec.js [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    new User({userId: 's03134054', password: '123456'}).save(function (err, data) {
+      if (err) return done.fail(err);
+      request(app)
+        .post('/api/sessions')
+        .send({userId: 's03134054', password: '123456'})
+        .expect({httpCode: 201, message: "SUCCESS", newUser: false}, finish(done))
```
Comment (freewind) [inline spec/sessions-spec.js:18]: .expect(201, {message, ....} , finish(done)

File: spec/sessions-spec.js [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+    })
+  });
+
+  it('post empty id and password 1 sessions', (done) => {
```
Comment (freewind) [inline spec/sessions-spec.js:22]: 中文

Comment (freewind) [inline spec/sessions-spec.js:22]: describe('bad request', ...

File: spec/sessions-spec.js [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+      .post('/api/sessions')
+      .send({userId: ''})
+      .expect({httpCode: 400, message: "数据不能为空"}, finish(done))
+  });
+  it('post error id and password sessions', (done) => {
+    request(app)
+      .post('/api/sessions')
+      .send({userId: 's03134054', password: '122456'})
+      .expect({httpCode: 401, message: "用户名或密码有误，登录失败"}, finish(done))
+  });
```
Comment (freewind) [inline spec/sessions-spec.js:45]: 第一次登录的情况没有测试

**Pull request merged**